tree-sitter-stack-graphs/src/lib.rs

```diff
@@ -550,52 +632,51 @@ impl From<tree_sitter_graph::ExecutionError> for LoadError {
     }
 }
 
-struct StackGraphLoader<'a> {
-    stack_graph: &'a mut StackGraph,
-    file: Handle<File>,
-    graph: &'a Graph<'a>,
-    source: &'a str,
-    span_calculator: SpanCalculator<'a>,
-}
+impl<'a> Builder<'a> {
+    fn load(mut self, cancellation_flag: &dyn CancellationFlag) -> Result<(), LoadError> {
+        let cancellation_flag: &dyn stack_graphs::CancellationFlag = &cancellation_flag;
 
-impl<'a> StackGraphLoader<'a> {
-    fn new(
-        stack_graph: &'a mut StackGraph,
-        file: Handle<File>,
-        graph: &'a Graph<'a>,
-        source: &'a str,
-    ) -> Self {
-        let span_calculator = SpanCalculator::new(source);
-        StackGraphLoader {
-            stack_graph,
-            file,
-            graph,
-            source,
-            span_calculator,
+        // By default graph ids are used for stack graph local_ids. A remapping is computed
+        // for local_ids that already exist in the graph---all other graph ids are mapped to
+        // the same local_id.
+        let mut next_local_id = (self.graph.node_count() - self.injected_node_count) as u32;
+        for node in self.stack_graph.nodes_for_file(self.file) {
+            let local_id = self.stack_graph[node].id().local_id();
+            let index = (local_id as usize) + self.injected_node_count;
+            // find next available id for which no stack graph node exists yet
+            while self
+                .stack_graph
+                .node_for_id(NodeID::new_in_file(self.file, next_local_id))
+                .is_some()
+            {
+                next_local_id += 1;
+            }
+            // remap occupied local_id to next available id
+            self.remapped_local_ids
+                .insert(index, NodeID::new_in_file(self.file, next_local_id));
         }
-    }
-}
-
-impl<'a> StackGraphLoader<'a> {
-    fn load(&mut self, cancellation_flag: &dyn CancellationFlag) -> Result<(), LoadError> {
-        let cancellation_flag: &dyn stack_graphs::CancellationFlag = &cancellation_flag;
 
         // First create a stack graph node for each TSG node.  (The skip(2) is because the first
         // two DSL nodes that we create are the proxies for the stack graph's “root” and “jump to
         // scope” nodes.)
-        for node_ref in self.graph.iter_nodes().skip(2) {
+        let injected_node_count = self.injected_node_count;
+        for node_ref in self
+            .graph
+            .iter_nodes()
+            .filter(|n| n.index() >= injected_node_count)
+        {
             cancellation_flag.check("loading graph nodes")?;
-            let node = &self.graph[node_ref];
```

Comment:
Passing the node directly didn't play well with the borrow checker. The node held a borrow of `self.graph`, which would block the methods calls which borrow `&mut self`.